File: Problems/0643_Maximum_Average_Subarray1/Project_Python3/Maximum_Average_Subarray1.py
import itertools
import os
import operator
import sys
import time

class Solution:
    def findMaxAverage2(self, nums: 'List[int]', k: 'int') -> 'float':
        sums = [0] + list(itertools.accumulate(nums))
        return max(map(operator.sub, sums[k:], sums)) / k

    def findMaxAverage(self, nums, k):
        current_mean = sum(nums[:k])
        max_mean = current_mean
        
        for i in range(len(nums) - k):
            current_mean -= nums[i]
            current_mean += nums[i + k]
            
            if current_mean > max_mean:
                max_mean = current_mean
            
        return max_mean / k

def str_to_int_array(flds):
    if len(flds) <= 0:
        return None
    temp = flds.split(",")
    nums = [0]*len(temp)
    for i in range(len(temp)):
        nums[i] = int(temp[i])
    return nums

def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("argv[1] = %s" %temp)
        loop_main(temp)
    #   print("Hit Return to continue...")
    #   input()

def loop_main(temp):
    str_args = temp.replace("\"","").replace("[[","").replace("]]","").rstrip().split("],[")
    nums = str_to_int_array(str_args[0])
    k = int(str_args[1])
    print("nums[] = %s, k = %d" %(nums, k))

    time0 = time.time()

    sl = Solution()
    result = sl.findMaxAverage(nums, k)

    print("result = %s" %result)

    time1 = time.time()
    print("Execute time ... : %f[s]\n" %(time1 - time0))

if __name__ == "__main__":
    main()

File: Problems/0643_Maximum_Average_Subarray1/Project_CS/Maximum_Average_Subarray1.cs
using System;
using System.Collections.Generic;

public class Solution
{
    public double FindMaxAverage(int[] nums, int k)
    {
        int len = nums.Length;

        if (len == 1)
            return nums[0];

        double sum = 0;
        for(int x = 0; x < k; x++)
        {
            sum += nums[x];
        }

        double maxSum = sum;
        for(int i = 1; i <= len - k; i++)
        {
            sum = sum - nums[i - 1] + nums[i + k - 1];
            if (sum > maxSum)
                maxSum = sum;
        }
        
        return maxSum/k;
    }

    public int[] str_to_int_array(string s)
    {
        if (s.Length <= 0)
            return null;

        string[] flds = s.Split(',');
        int[] nums = new int[flds.Length];

        if (flds.Length <= 0)
            return nums;

        for (int i = 0; i < nums.Length; ++i)
        {
            nums[i] = int.Parse(flds[i]);
        }

        return nums;
    }

    public string output_int_array(int[] nums)
    {
        if (nums.Length <= 0)
            return "";

        string resultStr = "[" +  nums[0].ToString();
 
        for (int i = 1; i < nums.Length; ++i)
        {
            resultStr += "," + nums[i].ToString();
        }

        return resultStr + "]";
    }

    public void Main(string args)
    {
        string arg_str = args.Replace("[[","").Replace("]]","").Trim();
        string[] flds = arg_str.Split(new string[] {"],["}, StringSplitOptions.None);
        int[] nums = str_to_int_array(flds[0]);
        int k = int.Parse(flds[1]);

        Console.WriteLine("nums = " + output_int_array(nums) + ", k = " + k.ToString());
        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        double result = FindMaxAverage(nums, k);
        Console.WriteLine("result = " + result.ToString());
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}
